Convert `FastProcessor` to iterative execution

plafer opened the pull request:
Closes #1989

Note: `ContinuationStack` is meant to replace/be merged with the `ExecutionTraversal` in #1839.

Unfortunately, we get the same performance on the `blake3_1to1` benchmark with this change (where we were hoping for an improvement). I didn't try to optimize too much yet, as I'd focus on fully merging #1839 first.

bitwalker commented on processor/src/continuation_stack.rs:
```diff
@@ -0,0 +1,193 @@
+/// which gives the processor sufficient information to execute the node correctly. Note that
+/// exactly how "start" and "end" are used depends on the node type being processed.
+#[derive(Debug, Default)]
+pub enum ProcessingStepPhase {
```
IMO this can just be combined with `ProcessingStep`, and then the prologue/epilogue of each node type handler be split out into separate functions.

This will be much friendlier to the branch predictor, because the dispatch loop will only call a given handler for a specific `ProcessingStep` variant, which in turn will lead to more predictable control flow patterns, i.e. every time the `execute_join_node_start` function is called from the call site in the match arm of `ProcessingStep::Prologue`, the branch predictor can predict with great certainty that the next branch in the dispatch loop will be to the `ProcessingStep::Node` arm and from there to `execute_join_node` - and so on.

plafer replied:
Fixed

bitwalker commented on processor/src/continuation_stack.rs:
```diff
@@ -0,0 +1,193 @@
+/// This enum defines the different types of processing steps that can be performed on MAST nodes
+/// during program execution. Each variant represents a specific phase of node processing.
+#[derive(Debug)]
+pub enum ProcessingStep {
```
I find it a bit confusing and verbose to call this `ProcessingStep` versus `Continuation`, particularly given that this module is notionally a stack of  continuations, not processing steps. The fact that there a couple different types of continuations doesn't really change that IMO.

bitwalker reviewed (commented):
I took a look through with an eye to why performance hasn't improved, and I think I've identified a couple of issues that likely contribute, but even if they make no difference, may still simplify things in a couple ways.

bobbinth commented on processor/src/fast/mod.rs:
```diff
@@ -320,17 +323,86 @@ impl FastProcessor {
+                        },
+                    }
+                },
+                ProcessingStep::PreDecorators(node_id) => {
+                    let node = current_forest.get_node_by_id(node_id).unwrap();
+                    for &decorator_id in node.before_enter() {
+                        self.execute_decorator(&current_forest[decorator_id], 0, host)?;
+                    }
+                },
+                ProcessingStep::PostDecorators(node_id) => {
+                    let node = current_forest.get_node_by_id(node_id).unwrap();
+                    for &decorator_id in node.after_exit() {
+                        self.execute_decorator(&current_forest[decorator_id], 0, host)?;
+                    }
+                },
```
Related to the previous comment, I would probably handle the decorators in node processing functions. It will add a bit more boilerplate code, but should also simplify continuation stack processing - i.e., the `ProcessingStep` could look something like this:
```rust
pub enum ProcessingStep{
    Start(MastNodeId),
    Finish(MastNodeId),
}
```
We could also consider making the `Finish` variant more explicit (because we always know what type of a node we are finishing) - e.g.:
```rust
pub enum ProcessingStep{
    Start(MastNodeId),
    FinishJoin, // don't care about node ID
    FinishSplit, // don't care about node ID
    FinishLoop(MastNodeId),
    ...
}
```

plafer replied:
Done

bobbinth commented on processor/src/fast/mod.rs:
```diff
@@ -349,294 +421,315 @@ impl FastProcessor {
+                        continuation_stack.push_node_with_decorators(callee_id);
+                    },
+                    None => {
+                        let mast_forest =
+                            host.get_mast_forest(&callee_hash).await.ok_or_else(|| {
+                                ExecutionError::dynamic_node_not_found(callee_hash, &err_ctx)
+                            })?;
```
This doesn't seem to handle loading of the advice data from the new MAST forests (i.e., similar to how it is done in `resolve_external_node()` on line 1090 - 1092 below). We should probably encapsulate loading of MAST forests into a small helper procedure which is used both here and there.

Also, do we not need to call  `continuation_stack.push_new_forest()` here similar to how we do it on line 731 below?

plafer replied:
Fixed - although I didn't create the function that encapsulates yet (will give another look tomorrow). The (small) challenge is that we return different errors when loading a mast forest depending on if we're processing an external node or dyn node - so I just have to think how to make this cleanest.

plafer replied:
> This doesn't seem to handle loading of the advice data from the new MAST forests (i.e., similar to how it is done in resolve_external_node() on line 1090 - 1092 below).

This was also missing in `Process::execute_dyn_node()`, so I fixed it there too. I did it as a quick fix though, since we're planning on removing `Process` soon.

plafer replied:
> We should probably encapsulate loading of MAST forests into a small helper procedure which is used both here and there.

Now done too

bobbinth reviewed (commented):
Looks good! Thank you! Not a review yet, but I left a few comments inline.

plafer commented on processor/src/continuation_stack.rs:
```diff
@@ -0,0 +1,193 @@
+/// This allows the processor to execute a program iteratively in a loop rather than recursively
+/// traversing the nodes.
+pub struct ContinuationStack {
+    forest_continuation_stack: Vec<ForestContinuationStack>,
```
Fixed

plafer commented on processor/src/fast/mod.rs:
```diff
@@ -349,294 +421,315 @@ impl FastProcessor {
+        current_node_id: MastNodeId,
+        phase: ProcessingStepPhase,
+        continuation_stack: &mut ContinuationStack,
     ) -> Result<(), ExecutionError> {
```
fixed

plafer commented:
@bitwalker @bobbinth addressed the comments, but will need to do a bit of cleanup in the morning - it's in a decent state, but I haven't explicitly updated the docstrings. Feel free to give it another look, or wait for when it's fully cleaned up tomorrow - up to you.

For some reason the CI fails to build, but it works fine locally, so I'll look into that as well.

bitwalker commented on core/src/mast/mod.rs:
```diff
@@ -794,3 +794,27 @@ pub enum MastForestError {
+///
+/// Adapted from the `downcast!` macro in the `ibc-relayer-types` crate.
+#[macro_export]
+macro_rules! downcast_node {
```
I would suggest using the following pattern instead:

```rust
impl MastNode {
    #[inline]
    #[track_caller]
    pub fn unwrap_basic_block(&self) -> &BasicBlockNode {
        match self {
            Self::Block(node) => node,
            other => unexpected_node_type(other, "basic block"),
        }
    }
}

#[cold]
#[inline(never)]
#[track_caller]
fn unwrap_node_type_failed(node: &MastNode, expected: &str) -> ! {
    let actual = match node {
        MastNode::Block(_) => "basic block",
        MastNode::Join(_) => "join",
        ...
    };
    panic!("tried to unwrap {expected} node, but got {actual}");
}
```

A few reasons for this:

1. The `unwrap` terminology is conventional for this sort of thing. If you look at how `unwrap`/`expect` are implemented on types like `Option`/`Result`, they use this sort of approach. The `downcast` terminology is reserved specifically for casting trait objects to a more concrete type, e.g. `&dyn Any` to `&T`. When I initially saw this, I had to remind myself that `MastNode` wasn't a trait.
2. The actual panic is defined in a single place, but will still point to the actual place where the unwrap was attempted due to the `#[track_caller]` annotations, so a lot less code is generated.
3. The `#[cold]`/`#[inline(never)]` annotations on the panic path ensure that all of that code is outlined away from the hot path, hints to the compiler that the branch is rarely taken, and will keep it in a different page so that it isn't fetched into the cache with hot code paths.

plafer commented:
> Thank you! Did these changes affect the performance in any way?

Unfortunately not. My current hunch is that `ContinuationStack::pop_stack()` cloning the `Arc<MastForest>` (acquiring a mutex internally) is the main cause of performance degradation. But we are currently forced to do it for borrow-checker reasons. There's hopefully a way to re-design the API to avoid that, but I haven't spent enough time looking into it.

adr1anh commented on processor/src/fast/mod.rs:
```diff
@@ -320,17 +323,113 @@ impl FastProcessor {
-            .await?;
+        let mut continuation_stack = ContinuationStack::new(program);
+
+        while let Some((current_forest, processing_step)) = continuation_stack.pop_continuation() {
```
Would it make sense to extract this into an `execute_step` function? This was something I think was discussed async yesterday.

adr1anh commented on processor/src/fast/mod.rs:
```diff
@@ -963,31 +1175,29 @@ impl FastProcessor {
+    // HELPERS
+    // ----------------------------------------------------------------------------------------------
+
+    async fn load_mast_forest<E>(
```
Below we can add the `err_ctx` to the `advice_error`

plafer replied:
Good catch - fixed

adr1anh reviewed (approved):
Minor nits, but otherwise look really good!

bobbinth commented on processor/src/continuation_stack.rs:
```diff
@@ -0,0 +1,155 @@
+
+    /// Pops the next continuation from the continuation stack, and returns it along with its
+    /// associated MAST forest.
+    pub fn pop_continuation(&mut self) -> Option<(Arc<MastForest>, Continuation)> {
+        let continuation = self.stack.pop()?;
+        Some((self.current_forest.clone(), continuation))
+    }
```
Probably not for this PR, but I wonder if there is a way to avoid cloning `Arc` for MAST forest on every continuation pop.

plafer replied:
Went for it in this PR, and it made the `blake3_1to1` benchmark go from ~1.82-1.85ms down to ~1.79ms (still up from ~1.40ms before the change to async).

bobbinth commented on processor/src/fast/mod.rs:
```diff
@@ -320,17 +323,113 @@ impl FastProcessor {
+                        .await?
+                },
+                Continuation::StartJoin(node_id) => {
+                    let join_node = current_forest[node_id].unwrap_join();
+                    self.start_join_node(
+                        join_node,
+                        node_id,
+                        &current_forest,
+                        &mut continuation_stack,
+                        host,
+                    )?
```
nit: we could look up the `join_node` inside the `start_join_node()` function. Then, this would look something like:
```rust
self.start_join_node(node_id, &current_forest, &mut continuation_stack, host)?
```
(i.e., passing both `join_node` and `node_id` into the function is a bit redundant)

This also applies to other node types.

plafer replied:
Ended up reverting this change due to your [other suggestion](https://github.com/0xMiden/miden-vm/pull/1990#discussion_r2211939045) for most nodes (all except `Dyn` and `External` which there were borrow-checker issues), since it made reduced the number of `MastForest` accesses.

bobbinth commented on processor/src/continuation_stack.rs:
```diff
@@ -0,0 +1,155 @@
+///
+/// This enum defines the different types of continuations that can be performed on MAST nodes
+/// during program execution.
+#[derive(Debug)]
+pub enum Continuation {
```
Not for this PR: the current structure of this requires us to do a lookup into the MAST forest (via the `get_node_by_id()` function) twice for every node: first time when we push it into the continuation stack, and second time when we pop it.

Not sure if it will make practical difference, but we should be able to avoid this by removing node-specific `Start*` variants. It could look something like:
```rust
pub enum Continuation {
    StartNode(MastNodeId),
    FinishJoin(MastNodeId),
    FinishSplit(MastNodeId),
    FinishLoop(MastNodeId),
    FinishCall(MastNodeId),
    FinishDyn(MastNodeId),
    EnterForest(Arc<MastForest>),
}
```

plafer replied:
Ah yes good suggestion! Went for it in this PR, since it reduces the number of variants quite a bit.

bobbinth commented on processor/src/fast/mod.rs:
Not for this PR, but this file has gotten pretty big. We should split it up into 2 - 3 files in subsequent PRs.

bobbinth reviewed (approved):
Looks good! Thank you. I left a few comments inline, but most of them are for subsequent PRs.
